MAVProxy/modules/mavproxy_speech.py

```diff
@@ -68,50 +255,156 @@ def say_speechd(self, text, priority='important'):
+                announcements.dispatch(msg, self.say)
 
+    def print_and_say(self, text, priority='important'):
+        self.console.writeln(text)
+        if self.enabled_for == "ALL":
+            self.say(text, priority)
+        elif self.enabled_for == "MODE" and len(text) > 4 and text[0:4] == "Mode":
+            self.say(text, priority)
+
+    def say(self, text, priority='important'):
+        '''speak some text'''
+        ''' http://cvs.freebsoft.org/doc/speechd/ssip.html see 4.3.1 for priorities'''
+        if self.settings.speech and self.say_backend is not None:
+            self.say_backend(text, priority=priority)
+
+    #
+    # Command line configuration commands.
+    # NOTE: Speech can be enabled or disabled and changed from main mpstate settings.
+    #
     def cmd_speech(self, args):
         '''speech commands'''
-        usage = "usage: speech <say>"
+        usage =  "usage: speech <say|list|enable|add|remove>\n"
+        usage += "  Call any command with no arguments for details"
         if len(args) < 1:
             print(usage)
             return
 
         if args[0] == "say":
+            cmd_usage  = "usage: speech say <text to say>\n"
+            cmd_usage += "  Say arbitrary text. Useful for testing"
             if len(args) < 2:
-                print("usage: speech say <text to say>")
+                print(cmd_usage)
                 return
             self.say(" ".join(args[1::]))
+        elif args[0] == "enable":
+            cmd_usage  = "usage: speech enable (ALL|MODE|MAV|NONE)\n"
+            cmd_usage += "  ALL  - reports all console and mavlink messages\n"
+            cmd_usage += "  MODE - reports only mode changes and mavlink messages\n"
+            cmd_usage += "  MAV  - reports only mavlink messages\n"
+            cmd_usage += "  NONE - disables speech"
+            if len(args) < 2:
+                print cmd_usage
+            else:
+                command = args[1].upper()
+                if (command in ["ALL", "MODE", "MAV", "NONE"]):
+                    self.enabled_for = command
+                    if args[1] == "NONE":
+                        self.settings.set('speech', 0)
+                    else:
+                        self.settings.set('speech', 1)
+                else:
+                    print cmd_usage
+        elif args[0] == "list":
+            print "Speech module will announce:", self.enabled_for
+            print "Speech module knows how to report: "
+            for idx, announcement in enumerate(self.msgs_to_announce):
+                print "#%d: %s" % (idx, announcement)
+        elif args[0] == "add":
+            usage =  "usage: speech\n"
+            usage += "add <msg_type>.<msg_entry> periodic    (every=<seconds>) (as=<say_as>) (scale=<scale>) (format=\"<formatstr>\")\n"
+            usage += "add <msg_type>.<msg_entry> conditional (every=<seconds>) (as=<say_as>) (scale=<scale>) (format=\"<formatstr>\") (max=<val>) (min=<val>)\n"
```

Comment:
If my suggestion for allowing announcements of expressions (based on whether an expressoin returns true....) is accepted, then conditional and onchange can probably be merged into one.  Old value and new value would need to be supplied to the expression, and some method for state storage (we already do something like this using `m[]`.

Doing this would allow reading out deltas rather than absolute values, for example.